### ISSUE TITLE
Read char16_t* String in user defined struct

### ISSUE THREAD

**Issue by araisch** [wlav/cppyy#4]
According to that [SO-Question](https://stackoverflow.com/questions/68710391/read-char16-t-string-in-user-defined-struct-with-cppyy-from-c/):

C++ Code:
```
#define MLPI_MOTION_MAX_AXIS_NAME_LEN        (80)
typedef unsigned int                  ULONG;
typedef char16_t                    WCHAR16;
typedef enum MlpiAxisType
{<snipp>
}MlpiAxisType;

typedef struct MlpiAxisInformation
{
  MlpiAxisRef     axis;                                 //!< Logical axis address.
  ULONG           deviceAddress;                        //!< (SERCOS) device address.
  MlpiAxisType    axisType;                             //!< Type of axis (virtual, real, etc...).
  WCHAR16         name[MLPI_MOTION_MAX_AXIS_NAME_LEN];  //!< The axis name.
}MlpiAxisInformation;

MLPI_API MLPIRESULT mlpiMotionGetConfiguredAxes(const MLPIHANDLE connection, MlpiAxisInformation* configAxes, const ULONG numElements, ULONG *numElementsRet);
```
Python-Code (working more or less):
```
class MlpiAxisRef(ctypes.Structure):
    _fields_ = [
        ("controlNo",ctypes.c_int16),
        ("axisNo",ctypes.c_int16) ]

class MlpiAxisInformation(ctypes.Structure):
    _fields_ = [
        ("axis", MlpiAxisRef),
        ("deviceAddress",ctypes.c_ulong),
        ("axisType", ctypes.c_int16),
        ("name", ctypes.c_uint16*79) ]

def MLPIGetConfiguredAxes(self) -> List[MlpiAxisInformation]:
        length = ctypes.c_ulong(99)
        length_ret = ctypes.c_ulong(0)
        self._mlpi.mlpiMotionGetConfiguredAxes.argtypes = (ctypes.c_ulonglong, ctypes.POINTER(MlpiAxisInformation), ctypes.c_ulong, ctypes.POINTER(ctypes.c_ulong))
        self._mlpi.mlpiMotionGetConfiguredAxes.restype = ctypes.c_long
        val = (MlpiAxisInformation*100)()
        ret = self._mlpi.mlpiMotionGetConfiguredAxes(self.con, val, length, length_ret)
<some code to covert int-array to string>
```
It works, but it's not fully correctly assigned. So not each byte is on the right position. Tried various datatypes and with `_pack_=1` and without. It changes, but somewhere there is always anything wrong. Therefore my idea was to use cppyy to get it running. 
- Is cppyy a better tool for that than ctypes?
- Is there a possibility to use the Types from the C++ Code and cast them in Python later? So tried to get a `cppyy.ll.array_new[cppyy.gbl.MlpiAxisInformation](length.value)`, but array_new just supports standard types?
- Or generate a wrapper function and return clean results to python? 

Edit: To be clear: I'm sure this could be solved with ctypes, but it is a pain to guess the right byte-size in this structs. And even if you guess right, on another platform it could be compiled differently. 
I'm asking if there is a way to achieve that in cppyy more convenient, with some sort of mechanism taking the Types from C++ and get the transfer to python code type safe.

**Comment by araisch** [comment]
Ok got the solution for the above function by pure guessing:
```
class MlpiAxisRef(ctypes.Structure):
    #_pack_=1
    _fields_ = [
        ("controlNo",ctypes.c_uint),
        ("axisNo",ctypes.c_uint) ]

class MlpiAxisInformation(ctypes.Structure):
    _pack_=1
    _fields_ = [
        ("axis", MlpiAxisRef),
        ("deviceAddress",ctypes.c_uint),
        ("axisType", ctypes.c_uint),
        ("name", ctypes.c_uint16*80) ]
```
Missed the ulong type def :(
But anyway: Is there a way to solve it in cppyy more convenient?

**Comment by wlav** [comment]
The enums should not be a problem and pick the correct type automatically, but there is a recent fix for class enums of a typedef'd type. Otherwise for creating an array of objects, I'd recommend `std::vector` (there is no optimization benefit to using `std::array`. The low-level stuff only exists for those cases where the other side will manage (and delete) the memory. (That said, it should have worked; the reshaping failing is a bug.)

Anyway, the above is still not a running example, so I have to make up some pieces. How does this look:

```import cppyy
import cppyy.ll

cppyy.cppdef(r"""\
struct AxisInformation {
    char16_t name[16];
};

void fillem(AxisInformation* a, int N) {
    char16_t fill[] = {u'h', u'e', u'l', u'l', u'o', u'\0'};
    for (int i = 0; i < N; ++i)
        memcpy(a[i].name, fill, sizeof(fill));
}""")

N = 10

va = cppyy.gbl.std.vector['AxisInformation'](N)
cppyy.gbl.fillem(va.data(), N)

for ai in va:
    print(cppyy.ll.cast['char16_t*'](ai.name))
```

**Comment by wlav** [comment]
This one had regressed; now fixed in repo (as is the reshaping issue, although `std::vector` remains recommended).

**Comment by wlav** [comment]
Fix released with 2.3.0 and its dependencies.
